File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarLength.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarLengthData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.LengthUnit;

/**
 * @author abray
 * Length Scalar Class, will enforce that units are proper to Lengths
 */

public class SEScalarLength extends SEScalar
{
  public SEScalarLength()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarLength(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarLength(double value, LengthUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarLengthData src, SEScalarLength dst)
  {
    SEScalar.load(src.getScalarLength(),dst);
  }
  public static ScalarLengthData unload(SEScalarLength src)
  {
    if(!src.isValid())
      return null;
    ScalarLengthData.Builder dst = ScalarLengthData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarLength src, ScalarLengthData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarLengthBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, LengthUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(LengthUnit unit)
  {
    return this.getValue(unit.toString());
  }
  
  public boolean validUnit(String unit)
  {
    if(LengthUnit.validUnit(unit))
      return true;
    return false;
  }
}

File: cdm/cpp/properties/SEScalarLength.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarLength.h"

const LengthUnit LengthUnit::m("m");
const LengthUnit LengthUnit::cm("cm");
const LengthUnit LengthUnit::mm("mm");
const LengthUnit LengthUnit::um("um");
const LengthUnit LengthUnit::in("in");
const LengthUnit LengthUnit::ft("ft");

bool LengthUnit::IsValidUnit(const std::string& unit)
{
  if (m.GetString().compare(unit) == 0)
    return true;
  if (cm.GetString().compare(unit) == 0)
    return true;
  if (mm.GetString().compare(unit) == 0)
    return true;
  if (um.GetString().compare(unit) == 0)
    return true;
  if (in.GetString().compare(unit) == 0)
    return true;
  if (ft.GetString().compare(unit) == 0)
    return true;
  return false;
}

const LengthUnit& LengthUnit::GetCompoundUnit(const std::string& unit)
{
  if (m.GetString().compare(unit) == 0)
    return m;
  if (cm.GetString().compare(unit) == 0)
    return cm;
  if (mm.GetString().compare(unit) == 0)
    return mm;
  if (um.GetString().compare(unit) == 0)
    return um;
  if (in.GetString().compare(unit) == 0)
    return in;
  if (ft.GetString().compare(unit) == 0)
    return ft;
  std::stringstream err;
  err << unit << " is not a valid Length unit";
  throw CommonDataModelException(err.str());
}

void SEScalarLength::Load(const cdm::ScalarLengthData& src, SEScalarLength& dst)
{
  SEScalarLength::Serialize(src, dst);
}
void SEScalarLength::Serialize(const cdm::ScalarLengthData& src, SEScalarLength& dst)
{
  SEScalarQuantity<LengthUnit>::Serialize(src.scalarlength(), dst);
}

cdm::ScalarLengthData* SEScalarLength::Unload(const SEScalarLength& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarLengthData* dst = new cdm::ScalarLengthData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarLength::Serialize(const SEScalarLength& src, cdm::ScalarLengthData& dst)
{
  SEScalarQuantity<LengthUnit>::Serialize(src, *dst.mutable_scalarlength());
}